### PR TITLE
many: implement our own ANSI-escape-using progress indicator

### PR DESCRIPTION
**Pull request by chipaca**
This introduces a `progress.Meter` written from "scratch" (mostly based on previous work and experiments I did on IO progress bars and fixed-width displays). It's called `progress.ANSIMeter`.

We were using a `progress.NullProgress` in situations where we actually didn't want progress bars but still wanted notifications, so `progress.QuietMeter` is a `progress.Meter` that does that, and that is returned by `progress.MakeProgressBar()` when it doesn't think we have an actual terminal.

We were replacing `os.Stdout` to look at the output of progress bars, so I wrote `progresstest.Meter` that just records what it's told instead of printing it anywhere, and use it in some places.

I renamed `progress.NullProgress` to `progress.NullMeter`, and added `progress.Null` as a default `progress.NullMeter`; as it's a `struct{}`, it doesn't make sense not to use that (we were using `&progress.NullMeter{}` in many many tests).

Lastly, `image` was not calling `Finished()` on the progress bar, so I added code for it to do that both in the sane case and in the case the user `^C`'s out.

### CODE REVIEW

File: progress/quantity.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+const (
+	period = 365.25 // julian years (c.f. the actual orbital period, 365.256363004d)
+)
+
+func divmod(a, b float64) (q, r float64) {
+	q = math.Floor(a / b)
+	return q, a - q*b
+}
+
+func formatDuration(dt float64) string {
```
Comment (zyga) [inline progress/quantity.go]: It would be nice to indicate that `dt` is in seconds.

File: progress/quantity.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	if dt < 60 {
+		if dt >= 9.995 {
+			return fmt.Sprintf("%.1fs", dt)
+		} else if dt >= .9995 {
+			return fmt.Sprintf("%.2fs", dt)
+		}
+
+		var prefix rune
+		for _, prefix = range "mun" {
```
Comment (zyga) [inline progress/quantity.go]: Just an idea, use:`"mµn"`

File: progress/quantity.go (hunk 1 above)
Comment (zyga) [inline progress/quantity.go]: Is there really no nice built-in library function to do all of this?

File: progress/quantity.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		if dt > 9.5 {
+			return fmt.Sprintf("%3.f%cs", dt, prefix)
+		}
+
+		return fmt.Sprintf("%.1f%cs", dt, prefix)
+	}
+
+	if dt < 600 {
+		m, s := divmod(dt, 60)
+		return fmt.Sprintf("%.fm%02.fs", m, s)
```
Comment (zyga) [inline progress/quantity.go]: The letters `h`, `m` and `s` denoting hours, minutes and seconds should be translatable.

File: progress/quantity.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		if d, h := divmod(dt, 24); h < 10 {
+			return fmt.Sprintf("%.fd%.fh", d, h)
+		}
+		return fmt.Sprintf("%4.1fd", dt/24)
+	}
+
+	dt /= 24 // dt now days
+
+	if dt < 2*period {
+		return fmt.Sprintf("%4.0fd", dt)
```
Comment (zyga) [inline progress/quantity.go]: Same comment about translatability of `d` and `y`

File: vendor/vendor.json [hunk 5]
```diff
@@ -151,12 +151,6 @@
 			"revisionTime": "2016-12-08T18:13:25Z"
 		},
 		{
-			"checksumSHA1": "qZxtczIIW7+bjlZ7QbfUJg7RI+s=",
```
Comment (zyga) [inline vendor/vendor.json]: I'm happy to see one go :)

File: progress/ansimeter.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	p.t0 = time.Now().UTC()
+	fmt.Print(cursorInvisible)
+}
+
+func norm(col int, msg []rune) []rune {
+	out := make([]rune, col)
+	copy(out, msg)
+	d := col - len(msg)
+	if d < 0 {
+		out[col-1] = '…'
```
[the remark this answers is not in the document]
Comment (chipaca) [reply progress/ansimeter.go]: In developing it I have checked gnome-terminal, xterm, rxvt, and the linux console. It works identically in all those places.

File: progress/ansimeter.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	}
+	fmt.Print("\r", enterReverseMode, string(msg[:i]), exitAttributeMode, string(msg[i:]))
+}
+
+func (p *ANSIMeter) Spin(msgstr string) {
+	col := getCol()
+	msg := norm(col, []rune(msgstr))
+	p.spin++
+	if p.spin > 90 {
+		p.spin = -p.spin + 1
```
[the remark this answers is not in the document]
Comment (chipaca) [reply progress/ansimeter.go]: a demo of the spinner

[the remark this answers is not in the document]
Comment (chipaca) [reply progress/ansimeter.go]: The spinner is 10% wide; when it's at 90%, it changes direction.

File: progress/quantity.go (hunk 2 above)
Comment (chipaca) [reply progress/quantity.go]: Yes, I might do that. In my library I didn't because it'd be the only non-ascii bit of it, and while I don't promise things to be ascii --- in fact I say
```golang
// The output string will be ascii right now, but this might change
// (there are some cases where using vulgar fractions would use the
// space a little better, e.g. FormatAmount(1500, 3) -> "1½k" instead
// of the current " 2k").
```
but this doesn't apply to this copy of the code obviously :-D

File: progress/quantity.go (hunk 1 above)
Comment (chipaca) [reply progress/quantity.go]: not so fixated on it being fixed-width while still conveying as much info as possible, no

File: progress/quantity.go (hunk 3 above)
Comment (chipaca) [reply progress/quantity.go]: I agree. OTOH the old one didn't AFAIK so maybe we can delay that

Comment (chipaca) [reply progress/quantity.go]: (just to confirm: cheggaaa's pb just does prints a time.Duration, which isn't localized)

Comment (chipaca) [reply progress/quantity.go]: one question i'd have is whether there's always a single-letter equivalent. E.g. in polish would you actually write `1g27m`?

Comment (zyga) [reply progress/quantity.go]: I think it's fair to treat this as string. Does chinese stuff fit into rune?

File: progress/quantity.go (hunk 2 above)
Comment (zyga) [reply progress/quantity.go]: But then you also use "..." fancy three-dot-unicode symbol :)

File: progress/ansimeter.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (zyga) [reply progress/ansimeter.go]: OOOH, that's one nice explanation!

File: progress/ansimeter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zyga) [reply progress/ansimeter.go]: What does it display in place of '…'?

File: progress/quantity.go (hunk 2 above)
Comment (chipaca) [reply progress/quantity.go]: Yeah, I'm going with `mµn`

File: progress/ansimeter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (chipaca) [reply progress/ansimeter.go]: dot dot dot

File: progress/quantity.go (hunk 3 above)
Comment (chipaca) [reply progress/quantity.go]: I'd have to ask somebody who spoke Mandarin. Unless what you mean is whether 时 is a rune in which case the answer is yes.

File: vendor/vendor.json (hunk 5 above)
Comment (zyga) [reply vendor/vendor.json]: Woah, linux cosole has advanced :D

File: progress/quantity.go (hunk 2 above)
Comment (zyga) [reply progress/quantity.go]: 😍 thanks

File: progress/quantity.go (hunk 3 above)
Comment (zyga) [reply progress/quantity.go]: I think we are fine then

File: progress/progress.go [hunk 8]
```diff
@@ -53,147 +51,55 @@ type Meter interface {
...
-// Start starts showing progress
-func (t *TextProgress) Start(label string, total float64) {
-	t.pbar = pb.New64(int64(total))
-	t.pbar.ShowSpeed = true
-	t.pbar.Units = pb.U_BYTES
-	t.pbar.Prefix(label)
-	t.pbar.Start()
-}
+// Null is a default NullMeter instance
+var Null = NullMeter{}
```
Comment (mvo5) [inline progress/progress.go]: :heart: I like that a lot!

File: image/helpers.go [hunk 9]
```diff
@@ -142,11 +144,24 @@ func (tsto *ToolingStore) DownloadSnap(name string, revision snap.Revision, opts
 	baseName := filepath.Base(snap.MountFile())
 	targetFn = filepath.Join(targetDir, baseName)
 
-	pb := progress.NewTextProgress()
+	pb := progress.MakeProgressBar()
+	defer pb.Finished()
+
+	// Intercept sigint
```
Comment (mvo5) [inline image/helpers.go]: Thanks for this!

File: image/helpers.go [hunk 10]
```diff
@@ -142,11 +144,24 @@ func (tsto *ToolingStore) DownloadSnap(name string, revision snap.Revision, opts
...
+	c := make(chan os.Signal, 3)
+	signal.Notify(c, syscall.SIGINT)
+	go func() {
```
Comment (mvo5) [inline image/helpers.go]: Should we have the "signal.Reset(...)" from below as a defer here?

File: progress/ansimeter.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+		}
+	}
+
+	msg := make([]rune, 0, col)
+	msg = append(msg, norm(col-len(percent)-len(speed)-len(timeleft), p.label)...)
+	msg = append(msg, []rune(percent)...)
+	msg = append(msg, []rune(speed)...)
+	msg = append(msg, []rune(timeleft)...)
+	i := int(current * float64(col) / p.total)
+	// the following can't happen... except when they do
```
Comment (mvo5) [inline progress/ansimeter.go]: *cough* I'm too jet-lagged to understand this comment, sorry. Could you rephrase slightly please :) ?

File: progress/ansimeter.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+	j := 1 + (spin+33)*col/100
+	fmt.Print("\r", string(msg[:i]), enterReverseMode, string(msg[i:j]), exitAttributeMode, string(msg[j:]))
+}
+
+func (*ANSIMeter) Finished() {
+	fmt.Print("\r", exitAttributeMode, cursorVisible, clrEOL)
+}
+
+func (*ANSIMeter) Notify(msgstr string) {
+	col := getCol()
```
Comment (mvo5) [inline progress/ansimeter.go]: (nitpick^2) I did not want to mention it, because its so tiny but reading it again and again makes me wonder if: `colWidth()` or something might be a better name for the helper?

File: progress/example_test.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/snapcore/snapd/progress"
+)
+
+func ExampleFormatAmount_short() {
```
Comment (mvo5) [inline progress/example_test.go]: Nice!

File: progress/progress.go [hunk 14]
```diff
@@ -53,147 +51,55 @@ type Meter interface {
...
+// MakeProgressBar creates an appropriate progress.Meter for the environ in
+// which it is called:
+//
+// * if MockMeter has been called, return that.
+// * if no terminal is attached, or we think we're running a test, a
+//   minimalistic QuietMeter is returned.
+// * otherwise, an ANSIMeter is returned.
+//
+// TODO: instead of making the pivot at creation time, do it at every call ( ....)
+var MakeProgressBar = func() Meter {
```
Comment (mvo5) [inline progress/progress.go]: I wonder if we need it as a var? It seems like we not actually replace `MakeProgressBar()` itself, there is the testing checks and the testMeter for this. Or was I not careful enough reading this?

File: progress/quantity.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+func formatBPS(n, sec float64, width int) string {
+	if sec < 0 {
+		sec = -sec
+	}
+	return formatAmount(uint64(n/sec), width-2) + "B/s"
+}
+
+const (
+	period = 365.25 // julian years (c.f. the actual orbital period, 365.256363004d)
```
Comment (mvo5) [inline progress/quantity.go]: :trophy: :-D

File: progress/quantity.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+		return fmt.Sprintf("%4.0f%s", dt, days)
+	}
+
+	dt /= period // dt now years
+
+	if dt < 9.995 {
+		return fmt.Sprintf("%4.2f%s", dt, years)
+	}
+
+	if dt < 99.95 {
```
Comment (mvo5) [inline progress/quantity.go]: I wonder if this code should have extra durations like millennia, epochs(sic), galactic year etc ;)

File: progress/progress.go [hunk 17]
```diff
@@ -53,147 +51,55 @@ type Meter interface {
...
+	if testMeter != nil {
+		return testMeter
 	}
-	r, _, err := reader.ReadRune()
-	if err != nil {
-		return false
+	if !inTesting && terminal.IsTerminal(int(os.Stdin.Fd())) {
```
Comment (mvo5) [inline progress/progress.go]: Unrelated to this branch - but I wonder if we should have something `osutil.OnTerminal()` that automatically adds the zero fd. Anyway, not important here. We have at least one `Interactive: terminal.IsTerminal(0)` in the code already.

File: progress/ansimeter.go [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+		return "---%"
+	}
+	q := p.written * 100 / p.total
+	if q > 999.4 || q < 0. {
+		return "???%"
+	}
+	return fmt.Sprintf("%3.0f%%", q)
+}
+
+func (p *ANSIMeter) Set(current float64) {
```
Comment (mvo5) [inline progress/ansimeter.go]: I wonder if we should have some more tests for ansimeter? It seems like it not having many tests currently.

File: progress/progress.go (hunk 14 above)
Comment (chipaca) [reply progress/progress.go]: I'll take a look, but it's quite possible it no longer needs to be a variable at all, which would make me happy.

File: progress/progress.go (hunk 17 above)
Comment (chipaca) [reply progress/progress.go]: we had `IsaTTY`, but dropped it in favour of this approach (because it used cgo, whereas this uses the syscall, but the syscall IIRC has arch shenanigans)

File: image/helpers.go (hunk 10 above)
Comment (chipaca) [reply image/helpers.go]: if image is expected to have any use beyond the one-shot interactive use it currently has then the goroutine needs cleanup and the signal reset, indeed. Otherwise nah :-)

File: progress/ansimeter.go (hunk 18 above)
Comment (chipaca) [reply progress/ansimeter.go]: :scream: no tests you say?!?

darn, you're probably right.

File: progress/ansimeter.go (hunk 11 above)
Comment (chipaca) [reply progress/ansimeter.go]: well, the count can't ever ever be bigger than the total, right? because nobody would ever put in a total that's wrong, right? Nor use a negative value to make it go backwards?

Right?

File: progress/ansimeter.go (hunk 12 above)
Comment (chipaca) [reply progress/ansimeter.go]: Names are my nemesis. Or my apotheosis. One of those two.

File: progress/quantity.go (hunk 16 above)
Comment (chipaca) [reply progress/quantity.go]: are there on-rune symbols for *any* of these? not afaik but let me know :-)

File: progress/ansimeter_test.go [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+
+type ansiSuite struct {
+	stdout *os.File
+}
+
+var _ = check.Suite(ansiSuite{})
+
+func (ansiSuite) TestNorm(c *check.C) {
+	msg := []rune(strings.Repeat("0123456789", 100))
+	high := []rune("🤗🤗🤗🤗🤗")
```
Comment (mvo5) [inline progress/ansimeter_test.go]: Hihi

File: progress/progress.go [hunk 20]
```diff
@@ -92,8 +92,8 @@ var inTesting bool = len(os.Args) > 0 && strings.HasSuffix(os.Args[0], ".test")
 //   minimalistic QuietMeter is returned.
 // * otherwise, an ANSIMeter is returned.
 //
-// TODO: instead of making the pivot at creation time, do it at every call ( ....)
-var MakeProgressBar = func() Meter {
+// TODO: instead of making the pivot at creation time, do it at every call.
+func MakeProgressBar() Meter {
```
Comment (mvo5) [inline progress/progress.go]: \o/

File: progress/quantity.go [hunk 21]
```diff
@@ -189,5 +189,14 @@ func formatDuration(dt float64) string {
...
 	}
 
-	return fmt.Sprintf("%4.f%s", dt, years)
+	if dt < 999.5 {
+		return fmt.Sprintf("%4.f%s", dt, years)
+	}
+
+	if dt > math.MaxUint64 || uint64(dt) == 0 {
+		// TODO: figure out exactly what overflow causes the ==0
+		return "ages!"
```
Comment (mvo5) [inline progress/quantity.go]: :heart:
